reconfigure dock clients on strut partial change

acrisci opened the pull request:
Strut partial tells i3 where to reserve space for i3bar. It is affected
by the `position` bar config directive.

acrisci commented:
I'm trying to fix #1294

acrisci commented:
This actually requires #1453 to work.

acrisci commented:
Now the bar position will change when we `reload`.

stapelberg commented:
I’ll review this once #1453 is in.

stapelberg commented on i3bar/src/xcb.c:
```diff
@@ -1462,6 +1462,51 @@ void realloc_sl_buffer(void) {
+            32,
+            12,
+            &strut_partial);
+
```
nit: stray newline

stapelberg commented on src/handlers.c:
```diff
@@ -1134,6 +1134,76 @@ static bool handle_class_change(void *data, xcb_connection_t *conn, uint8_t stat
+ * Handles the _NET_WM_STRUT_PARTIAL property for allocating space for dock clients.
+ *
+ */
+static bool handle_strut_partial_change(void *data, xcb_connection_t *conn, uint8_t state, xcb_window_t window,
```
Can you add a testcase for this please?

acrisci commented:
I added a test, but it is failing. However, it seems to work fine under normal conditions with i3bar.

The issue is that the request for the strut partial property fails with error code 3 (which I think is BadWindow). However, the con is found in the tree and the window id seems to be valid.

Any ideas on what is going on with the test?

stapelberg commented on src/handlers.c:
```diff
@@ -1174,6 +1174,85 @@ static bool handle_class_change(void *data, xcb_connection_t *conn, uint8_t stat
+    }
+
+    Con *search_at = croot;
+    Output *output = get_output_containing(con->geometry.x, con->geometry.y);
```
Using `con_get_output()` seems more straight-forward than `get_output_containing()` here.

stapelberg commented on src/handlers.c:
```diff
@@ -1174,6 +1174,85 @@ static bool handle_class_change(void *data, xcb_connection_t *conn, uint8_t stat
+    }
+
+    /* find the dockarea */
+    Con *dockarea = con_for_window(search_at, con->window, NULL);
```
`con_for_window` could return NULL. If you’re sure that it cannot under these circumstances, at least add an `assert`.

acrisci replied:
I think it should never be NULL because there is always a dock area below the output (or root) for the dock client.

stapelberg commented on testcases/t/102-dock.t:
```diff
@@ -143,6 +143,21 @@ wait_for_map $window;
+    pack('L12', 0, 0, 0, 16, 0, 0, 0, 0, 0, 0, 1280, 0)
+);
+
+@docked = get_dock_clients('bottom');
```
You’re missing a `sync_with_i3;` before this line.

stapelberg commented:
Even with the `sync_with_i3`, the testcase makes i3 crash sometimes. Valgrind reports invalid writes. I think your new code (or perhaps old code, but that’s rather unlikely) is corrupting memory somewhere.

acrisci commented:
It works just fine outside of the test environment.

stapelberg commented:
So? Memory corruption tends to have that feature (“works just fine on my machine”).

Commit:
i3bar: reconfigure strut partial on reload

Strut partial tells i3 where to reserve space for i3bar. It is affected
by the `position` bar config directive.

acrisci commented:
Thanks for the hint about the sync in the testcases. That led me to the memory error and I fixed it.

I messed up a rebase. Do I need to close this and submit another pull request?

Commit:
Handle strut partial changes for dock clients

Handle changes to the property _NET_WM_STRUT_PARTIAL by reallocating
screen space in the dock area when the dock client requests a new
position.

This will allow changes to the bar config directive `position` to be
effective on the `reload` command.

fixes #1294

stapelberg commented:
Not sure what you mean with the messed up rebase? Can you not use interactive rebase and then just `git push -f`?

acrisci commented:
It was showing a lot of recreated commits in the pull request for some reason. Maybe it was a github bug. I think that issue is resolved though and this should be ready.
